### ISSUE TITLE
Add a way to (optinally) export typenames

### ISSUE THREAD

**Issue by lenscas** [amethyst/rlua#195]
There are multiple projects to allow you to write lua with static types. One being https://github.com/teal-language/tl and another being https://github.com/TypeScriptToLua/TypeScriptToLua

I think it would be nice if there was a way for users to set names for their types, that way other projects can more easily be created that create the teal/typescript/whatever types for the rust objects.

I think the only changes that are needed is a new method in UserData that returns the name and one for ToLuaMulti and FromLuaMulti which gives an Vec/Iterator of names. Because AddMethods is already generic it should be possible for other crates to be created that use the names and the current AddMethods function to generate type definitions.

I don't mind opening an PR and I also understand if this is out of scope. However, I can't think of a nice way for crates to get these typenames without these changes.

Edit: Found a way to get the type information that is easy enough to use. Having those methods would still make it nicer though.

**Comment by jugglerchris** [comment]
Hi,
I'm not sure I completely understand what you need - can you summarise what you did to get type information please?
Thanks!

**Comment by lenscas** [comment]
I ended up making some wrapper structs and traits to make this all work. 

First I made a new trait that can be used to ask types how they should be called in lua: https://github.com/lenscas/tealr/blob/01ebb25b040eb3fb8ae611851d2d37d808b0cb9b/tealr/src/teal_data.rs#L31

I then made new traits that replace the use of UserData and UserDataMethods to enforce the use of the above trait
https://github.com/lenscas/tealr/blob/01ebb25b040eb3fb8ae611851d2d37d808b0cb9b/tealr/src/teal_data_methods.rs#L12
/blob/01ebb25b040eb3fb8ae611851d2d37d808b0cb9b/tealr/src/teal_data.rs#L4

and lastly, I made a wrapper struct and derive macros so that UserData can easily be implemented with my version of it https://github.com/lenscas/tealr/blob/01ebb25b040eb3fb8ae611851d2d37d808b0cb9b/tealr/src/user_data_wrapper.rs#L12

Basically, all I needed was a method on UserData (or as a separate trait that can be derived) so people can give their types a name which other libraries (like tealr) can then use to generate the correct type definition files.

**Comment by jugglerchris** [comment]
Hi,
It sounds like you've solved your problems, so I am closing this issue.  If I've misunderstood feel free to reopen.
Thanks!
